File: tests/render/render.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// The render tests' machinery: build a case, serve it, compare it.
//
// A case at a width is three steps, each a tool in tools/pageless-pdf:
//  1. pageless.ts compiles the document to a pageless PDF – one page as tall
//     as the document, every glyph where TeX put it – at the document's own
//     width plus `extra` pt;
//  2. site-from-run.ts builds a viewer page from the run at the document's own
//     width (once per case: every width is shown by the same page);
//  3. vector-compare.ts opens the page in Chromium with its column pinned to
//     the PDF's width and matches each glyph the viewer drew with the PDF's.
//
// Builds go to tests/render/build/<case>/ (ignored by git).
import { appendFileSync, createReadStream, existsSync, mkdirSync, readdirSync, rmSync, statSync, writeFileSync } from 'node:fs';
import { createServer, type Server } from 'node:http';
import { extname, join, normalize, relative, resolve, sep } from 'node:path';
import { fileURLToPath } from 'node:url';
import { pageless } from '../../tools/pageless-pdf/pageless.ts';
import { siteFromRun } from '../../tools/pageless-pdf/site-from-run.ts';
import { vectorCompare } from '../../tools/pageless-pdf/vector-compare.ts';
import { cases as settings, defaults, type CaseSettings } from './cases.ts';

const HERE = fileURLToPath(new URL('.', import.meta.url));
const REPO = resolve(HERE, '../..');
export const BUILD = join(HERE, 'build');

export interface Case extends CaseSettings {
  name: string; file: string; widths: number[]; tolerance: number; passes: number;
  /** the target every case is to meet (defaults.tolerance) */
  target: number;
}

/** Every case with its settings: cases/*.tex, and cases.ts's entries. */
export function allCases(): Map<string, Case> {
  const names = new Set([...readdirSync(join(HERE, 'cases')).filter(f => f.endsWith('.tex')).map(f => f.slice(0, -4)).sort(), ...Object.keys(settings)]);
  return new Map([...names].map(name => {
    const s = settings[name] ?? {};
    return [name, { ...defaults, ...s, name, target: defaults.tolerance,
      file: s.file ? join(REPO, s.file) : join(HERE, 'cases', `${name}.tex`),
      template: s.template ? join(REPO, s.template) : undefined } as Case];
  }));
}

export const buildDir = (c: Case, extra: number) => join(BUILD, c.name, extra ? `w${extra > 0 ? '+' : ''}${extra}` : 'w0');

/** A tool's lines, into a log file beside what it built. */
const logTo = (file: string) => { writeFileSync(file, ''); return (s: string) => appendFileSync(file, `${s}\n`); };

async function pagelessRun(c: Case, extra: number): Promise<string> {
  const out = buildDir(c, extra);
  rmSync(out, { recursive: true, force: true });
  mkdirSync(out, { recursive: true });
  await pageless(c.file, { out, passes: c.passes, widthExtra: `${extra}pt`, template: c.template, log: logTo(join(out, 'pageless.log')) });
  return out;
}

const sites = new Map<string, Promise<string>>();
/** The viewer page of a case, from its run at its own width (built once). */
function site(c: Case): Promise<string> {
  if (!sites.has(c.name)) sites.set(c.name, (async () => {
    const base = buildDir(c, 0);
    if (!existsSync(join(base, 'pageless.pdf'))) await pagelessRun(c, 0);
    const out = join(base, 'site');
    rmSync(out, { recursive: true, force: true });
    await siteFromRun(base, out, { log: logTo(join(base, 'site.log')) });
    return out;
  })());
  return sites.get(c.name)!;
}

export interface Vector {
  hsize_pt: number; window_pt: number;
  glyphs: { strip: number; viewer: number; matched: number };
  matched: { y: number; text: string; dx: number; dy: number }[];
  rules: { y: number; off?: number; dx?: number; dy?: number; unmatched?: boolean }[];
  rules_missing: { y: number }[];
}

/** The PDF at this width against the case's page: vector.json. */
export async function compare(c: Case, extra: number, urlRoot: string): Promise<Vector> {
  const page = await site(c);
  const strip = extra ? await pagelessRun(c, extra) : buildDir(c, 0);
  const rel = relative(BUILD, page).split(sep).join('/');
  return await vectorCompare(strip, `${urlRoot}/${rel}/index.html`, { out: join(strip, 'vector'), lineTol: c.tolerance, log: logTo(join(strip, 'compare.log')) }) as unknown as Vector;
}

/** build/ over HTTP on a free port, for the run. */
export function serve(): Promise<{ url: string; server: Server }> {
  mkdirSync(BUILD, { recursive: true });
  const types: Record<string, string> = { '.html': 'text/html', '.js': 'text/javascript', '.css': 'text/css', '.json': 'application/json', '.otf': 'font/otf', '.svg': 'image/svg+xml' };
  const server = createServer((req, res) => {
    let file = normalize(join(BUILD, decodeURIComponent(new URL(req.url ?? '/', 'http://x').pathname)));
    if (!file.startsWith(BUILD)) { res.writeHead(403).end(); return; }
    if (existsSync(file) && statSync(file).isDirectory()) file = join(file, 'index.html');
    if (!existsSync(file)) { res.writeHead(404).end(); return; }
    res.writeHead(200, { 'content-type': types[extname(file)] ?? 'application/octet-stream' });
    createReadStream(file).pipe(res);
  });
  return new Promise(ok => server.listen(0, '127.0.0.1', () => {
    const a = server.address() as { port: number };
    ok({ url: `http://127.0.0.1:${a.port}`, server });
  }));
}

/** The largest offset (pt, across or down) of a matched glyph, and of a
 *  matched rule – the farthest one of its corners is from TeX's (null when the
 *  browser drew no rules). */
export function worst(v: Vector): [number, number | null] {
  const g = Math.max(0, ...v.matched.map(m => Math.max(Math.abs(m.dx), Math.abs(m.dy))));
  const rs = v.rules.filter(r => r.off !== undefined).map(r => r.off!);
  return [g, rs.length ? Math.max(...rs) : null];
}

const f3 = (x: number) => x.toFixed(3);
const signed = (x: number) => `${x >= 0 ? '+' : ''}${f3(x)}`;

/** What is wrong in a comparison, in words; empty when it passes. */
export function problems(v: Vector, tolerance: number, ruleTolerance?: number, rulesMissing = 0): string[] {
  const out: string[] = [];
  const g = v.glyphs;
  if (!g.strip) out.push('the PDF has no glyphs');
  if (g.viewer !== g.strip) out.push(`the browser drew ${g.viewer} glyphs, TeX ${g.strip}`);
  if (g.matched !== g.viewer) out.push(`${g.viewer - g.matched} of the browser's glyphs have no glyph in the PDF within ${v.window_pt} pt`);
  const off = [...v.matched].sort((a, b) => Math.max(Math.abs(b.dx), Math.abs(b.dy)) - Math.max(Math.abs(a.dx), Math.abs(a.dy)))
    .filter(m => Math.max(Math.abs(m.dx), Math.abs(m.dy)) > tolerance);
  if (off.length) out.push(`${off.length} glyphs off by more than ${tolerance} pt: ` +
    off.slice(0, 5).map(m => `${JSON.stringify(m.text)} at y ${m.y} off by (${signed(m.dx)}, ${signed(m.dy)})`).join(', '));
  const unmatched = v.rules.filter(r => r.unmatched);
  if (unmatched.length) out.push(`${unmatched.length} rules drawn by the browser are not in the PDF`);
  if (v.rules_missing.length > rulesMissing)
    out.push(`${v.rules_missing.length} rules in the PDF the browser did not draw (${rulesMissing} allowed): ` +
      v.rules_missing.slice(0, 5).map(r => `at y ${r.y}`).join(', '));
  const rt = ruleTolerance ?? tolerance;
  const rulesOff = v.rules.filter(r => r.off !== undefined && r.off > rt).sort((a, b) => b.off! - a.off!);
  if (rulesOff.length) out.push(`${rulesOff.length} rules off by more than ${rt} pt: ` +
    rulesOff.slice(0, 5).map(r => `at y ${r.y} (corners ${f3(r.off!)} off; centre ${signed(r.dx!)}, ${signed(r.dy!)})`).join(', '));
  return out;
}

File: tests/render/render.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""The render tests' machinery: build a case, serve it, compare it.

A case at a width is three steps, each a tool in tools/pageless-pdf:

  1. pageless.py compiles the document to a pageless PDF – one page as tall
     as the document, every glyph where TeX put it – at the document's own
     width plus `extra` pt;
  2. site_from_run.py builds a viewer page from the run at the document's
     own width (once per case: every width is shown by the same page);
  3. vector_compare.py opens the page in Chromium with its column pinned to
     the PDF's width and matches each glyph the viewer drew with the PDF's.

Builds go to tests/render/build/<case>/ (ignored by git).
"""
import functools, hashlib, http.server, json, os, shutil, subprocess, sys, threading, tomllib
from pathlib import Path

import pytest

HERE = Path(__file__).resolve().parent
REPO = HERE.parent.parent
TOOLS = REPO / 'tools' / 'pageless-pdf'
BUILD = HERE / 'build'
PY = sys.executable
# Each test's worst offsets, kept on pytest's config for the summary
# (conftest.py): a list of (test id, case, (glyph, rule)).
WORST = pytest.StashKey[list]()


def cases() -> dict:
    """Every case with its settings: cases/*.tex, and cases.toml's sections."""
    conf = tomllib.loads((HERE / 'cases.toml').read_text())
    defaults = conf.pop('defaults', {})
    found = {p.stem: {} for p in sorted((HERE / 'cases').glob('*.tex'))}
    for name, extra in conf.items():
        found.setdefault(name, {}).update(extra)
    out = {}
    for name, extra in found.items():
        c = {**defaults, **extra, 'name': name}
        # the global target; `tolerance` is the case's own ceiling
        c['target'] = defaults['tolerance']
        c['file'] = REPO / c['file'] if 'file' in c else HERE / 'cases' / f'{name}.tex'
        if 'template' in c:
            c['template'] = REPO / c['template']
        out[name] = c
    return out


def run(cmd: list, log: Path) -> None:
    """Run a tool, its output into `log`; on failure, fail with its tail. The
    browser is tests/render's own Playwright (npm ci there), unless
    PLAYWRIGHT_DIR names another."""
    env = dict(os.environ)
    if 'PLAYWRIGHT_DIR' not in env and (HERE / 'node_modules' / 'playwright').exists():
        env['PLAYWRIGHT_DIR'] = str(HERE)
    r = subprocess.run([str(a) for a in cmd], capture_output=True, text=True, cwd=REPO, env=env)
    log.write_text(r.stdout + r.stderr)
    if r.returncode != 0:
        tail = '\n'.join((r.stdout + r.stderr).strip().splitlines()[-25:])
        raise RuntimeError(f'{Path(cmd[1]).name} failed (see {log}):\n{tail}')


def build_dir(case: dict, extra: int) -> Path:
    return BUILD / case['name'] / (f'w{extra:+d}' if extra else 'w0')


def pageless(case: dict, extra: int) -> Path:
    out = build_dir(case, extra)
    shutil.rmtree(out, ignore_errors=True)
    out.mkdir(parents=True)
    cmd = [PY, TOOLS / 'pageless.py', case['file'], '-o', out, '--passes', case['passes'],
           f'--width-extra={extra}pt']
    if 'template' in case:
        cmd += ['--template', case['template']]
    run(cmd, out / 'pageless.log')
    return out


@functools.lru_cache(maxsize=None)
def site(name: str) -> Path:
    """The viewer page of a case, from its run at its own width (built once)."""
    case = cases()[name]
    base = build_dir(case, 0)
    if not (base / 'pageless.pdf').exists():
        pageless(case, 0)
    out = base / 'site'
    shutil.rmtree(out, ignore_errors=True)
    run([PY, TOOLS / 'site_from_run.py', base, out], base / 'site.log')
    return out


def compare(case: dict, extra: int, url_root: str) -> dict:
    """The PDF at this width against the case's page: vector.json."""
    strip = build_dir(case, extra) if extra else build_dir(case, 0)
    page = site(case['name'])
    if extra:
        pageless(case, extra)
    rel = page.relative_to(BUILD).as_posix()
    out = strip / 'vector'
    run([PY, TOOLS / 'vector_compare.py', strip, f'{url_root}/{rel}/index.html', '--out', out,
         '--line-tol', case['tolerance']], strip / 'compare.log')
    return json.loads((out / 'vector.json').read_text())


class Server:
    """build/ over HTTP on a free port, in a thread, for the session."""
    def __init__(self):
        BUILD.mkdir(exist_ok=True)
        handler = functools.partial(Quiet, directory=str(BUILD))
        self.httpd = http.server.ThreadingHTTPServer(('127.0.0.1', 0), handler)
        self.url = f'http://127.0.0.1:{self.httpd.server_address[1]}'
        threading.Thread(target=self.httpd.serve_forever, daemon=True).start()

    def close(self):
        self.httpd.shutdown()


class Quiet(http.server.SimpleHTTPRequestHandler):
    def log_message(self, *args):
        pass


def worst(v: dict) -> tuple[float, float | None]:
    """The largest offset (pt, across or down) of a matched glyph, and of a
    matched rule (None when the browser drew no rules)."""
    g = max((max(abs(m['dx']), abs(m['dy'])) for m in v['matched']), default=0.0)
    rs = [max(abs(r['dx']), abs(r['dy'])) for r in v['rules'] if 'dx' in r]
    return g, (max(rs) if rs else None)


def problems(v: dict, tolerance: float, rule_tolerance: float | None = None) -> list[str]:
    """What is wrong in a comparison, in words; empty when it passes."""
    out = []
    g = v['glyphs']
    if not g['strip']:
        out.append('the PDF has no glyphs')
    if g['viewer'] != g['strip']:
        out.append(f"the browser drew {g['viewer']} glyphs, TeX {g['strip']}")
    if g['matched'] != g['viewer']:
        out.append(f"{g['viewer'] - g['matched']} of the browser's glyphs have no glyph in the PDF "
                   f"within {v['window_pt']} pt")
    worst = sorted(v['matched'], key=lambda m: -max(abs(m['dx']), abs(m['dy'])))
    off = [m for m in worst if max(abs(m['dx']), abs(m['dy'])) > tolerance]
    if off:
        sample = ', '.join(f"{m['text']!r} at y {m['y']} off by ({m['dx']:+.3f}, {m['dy']:+.3f})" for m in off[:5])
        out.append(f'{len(off)} glyphs off by more than {tolerance} pt: {sample}')
    unmatched_rules = [r for r in v['rules'] if r.get('unmatched')]
    if unmatched_rules:
        out.append(f'{len(unmatched_rules)} rules drawn by the browser are not in the PDF')
    rt = tolerance if rule_tolerance is None else rule_tolerance
    rules_off = sorted((r for r in v['rules'] if 'dx' in r and max(abs(r['dx']), abs(r['dy'])) > rt),
                       key=lambda r: -max(abs(r['dx']), abs(r['dy'])))
    if rules_off:
        sample = ', '.join(f"at y {r['y']} ({r['dx']:+.3f}, {r['dy']:+.3f})" for r in rules_off[:5])
        out.append(f'{len(rules_off)} rules off by more than {rt} pt: {sample}')
    return out
